Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,265 @@
+        // but the body itself is throwing
+        let result = await withTaskGroup(of: ChildTaskResult.self, returning: Result<Void, Error>.self) { group in
+            // First we have to register our signals.
+            let unixSignals = await UnixSignalsSequence(trapping: self.configuration.gracefulShutdownSignals)
+
+            group.addTask {
+                for await signal in unixSignals {
+                    return .signalCaught(signal)
+                }
+
+                return .signalSequenceFinished
+            }
+
+            // We have to create a graceful shutdown manager per service
+            // since we want to signal them individually and wait for a single service
+            // to finish before moving to the next one
+            var gracefulShutdownManagers = [GracefulShutdownManager]()
+            gracefulShutdownManagers.reserveCapacity(self.services.count)
+
+            for (index, service) in self.services.enumerated() {
+                self.logger.debug(
+                    "Starting service",
+                    metadata: [
+                        self.configuration.logging.serviceKey: "\(service)",
+                    ]
+                )
+
+                let gracefulShutdownManager = GracefulShutdownManager()
+                gracefulShutdownManagers.append(gracefulShutdownManager)
+
+                group.addTask {
+                    return await TaskLocals.$gracefulShutdownManager.withValue(gracefulShutdownManager) {
+                        do {
+                            try await service.run()
+                            return .serviceFinished(service: service, index: index)
+                        } catch {
+                            return .serviceThrew(service: service, index: index, error: error)
+                        }
+                    }
+                }
+            }
+
+            precondition(gracefulShutdownManagers.count == self.services.count, "We did not create a graceful shutdown manager per service")
+
+            // We are going to wait for any of the services to finish or
+            // the signal sequence to throw an error.
+            while !group.isEmpty {
+                // No child task is actually throwing here so the try! is safe
+                let result: ChildTaskResult? = await group.next()
+
+                switch result {
+                case .serviceFinished(let service, _):
+                    // If a long running service finishes early we treat this as an unexpected
+                    // early exit and have to cancel the rest of the services.
+                    self.logger.error(
+                        "Service finished unexpectedly. Cancelling all other services now",
+                        metadata: [
+                            self.configuration.logging.serviceKey: "\(service)",
+                        ]
+                    )
+
+                    group.cancelAll()
```

Comment:
Do we always want that? Especially the cancel? We might have unimportant services that, if they can't do their job anymore, we want to gracefully shutdown, but not cancel stuff that is in flight?

Reply:
This is basically coming back to the discussion of short vs long running services. After discussing this with many people I haven't heard a good example of a short running service that doesn't result int cancellation. If you have a service that is unimportant e.g. a service that sends logs out to some backend. Then you probably should model that service in a way that it just buffers logs and retries the connection.

In general, I would like to go forward without supporting short running services and revisit that topic later once we have concrete cases where we know we need this. Adding support for this can be done in a non-breaking change so I don't see it as a blocker.